### PR TITLE
Add SQLModelGenerator

### PR DESCRIPTION
**Pull request by KhirilovA**
## What?
I wrote another generator class that generates SQLModels, and saved a forcepk option ;)
## How?
I stayed with my idea to force primary key for the tables that don't have it, but only if you optional for DeclarativeGenerator and mandatory for SQLModelGenerator.

Based on DeclarativeGenerator i wrote a generator for SQLModels.
Columns are described with sa_column parameter

I didn't implement many-to-many relationships yet.

Screenshot_2

### CODE REVIEW

**Comment by agronholm** [comment]
I'm inclined to merge support for SQLModel, but I still don't feel comfortable with forcing primary keys.

By the way, I suggest you install the pre-commit git hook to prevent pre-commit.ci checks from failing here.

**Comment by KhirilovA** [comment]
I removed "forcepk" option from here and fixed pre-commit.ci checks from failing

**Comment by agronholm** [comment]
I will require new tests that cover this added functionality. When those tests are passing I will merge this.

**Comment by KhirilovA** [comment]
Ok, should i particapate in it?

**Comment by agronholm** [comment]
Any new feature needs to accompanied with tests, and yes, the author of said feature needs to add the tests.

**Comment by KhirilovA** [comment]
I added tests for checking index, constraint generation, one-to-one and one-to-many relationships

File: src/sqlacodegen/generators.py [hunk 1]
```diff
@@ -1342,3 +1342,132 @@ def render_relationship(self, relationship: RelationshipAttribute) -> str:
...
+            self.add_literal_import("sqlmodel", "Field")
+
+    def collect_imports_for_model(self, model: Model) -> None:
+        super(DeclarativeGenerator, self).collect_imports_for_model(model)
+        if isinstance(model, ModelClass):
+            for column_attr in model.columns:
+                if column_attr.column.nullable:
+                    self.add_literal_import("typing", "Optional")
+                    break
+            if model.relationships:
```
Comment (agronholm) [inline src/sqlacodegen/generators.py:1379]: Please add a blank line after each control block when more code follows on the same indentation level (e.g. to distinguish if...else blocks from unrelated code).
```suggestion

            if model.relationships:
```

File: src/sqlacodegen/generators.py [hunk 2]
```diff
@@ -1342,3 +1342,132 @@ def render_relationship(self, relationship: RelationshipAttribute) -> str:
...
+        except NotImplementedError:
+            self.add_literal_import("typing", "Any")
+        else:
+            self.add_import(python_type)
+
+    def render_module_variables(self, models: list[Model]) -> str:
+        declarations: list[str] = []
+        if any(not isinstance(model, ModelClass) for model in models):
+            declarations.append(f"metadata = {self.base_class_name}.metadata")
+        return "\n".join(declarations)
```
Comment (agronholm) [inline src/sqlacodegen/generators.py:1402]: Same as above. There are multiple such cases here, please change them all.
```suggestion

        return "\n".join(declarations)
```

File: tests/test_generators.py [hunk 3]
```diff
@@ -2545,3 +2546,157 @@ class Simple:
...
+        )
+        simple_items.indexes.add(Index("idx_number", simple_items.c.number))
+        simple_items.indexes.add(
+            Index("idx_text_number", simple_items.c.text, simple_items.c.number)
+        )
+        simple_items.indexes.add(Index("idx_text", simple_items.c.text, unique=True))
+
+        validate_code(
+            generator.generate(),
+            """\
```
Comment (agronholm) [inline tests/test_generators.py:2575]: Indent these text blocks the same way they're indented in existing code.

**Review by agronholm — CHANGES_REQUESTED** [review]
Looks promising, just need some code style adjustments.

File: src/sqlacodegen/generators.py [hunk 4]
```diff
@@ -1342,3 +1340,137 @@ def render_relationship(self, relationship: RelationshipAttribute) -> str:
...
+            return "".join(variables)
+
+        return ""
+
+    def render_column_attribute(self, column_attr: ColumnAttribute) -> str:
+        column = column_attr.column
+        try:
+            python_type = column.type.python_type
+        except NotImplementedError:
+            python_type_name = "Any"
```
Comment (agronholm) [inline src/sqlacodegen/generators.py:1431]: This needs the UUID column type hack from the dataclass generator.

**Comment by agronholm** [comment]
I suggest running the test suite locally using `tox` to determine if the changes you've made actually fix the test failures.

**Review by agronholm — APPROVED** [review]

**Comment by agronholm** [comment]
Thanks!

**Comment by KhirilovA** [comment]
Thank you for advices too!

**Comment by amacfie** [comment]
Cool. SQLModel is an available generator now? Should the README mention it?

**Comment by agronholm** [comment]
Yes, I have other documentation updates to make too before the next pre-release.

**Pull request merged**